### ISSUE TITLE
`DynetParams` object doesn't work any more.

### ISSUE THREAD

**Issue by xunzhang** [clab/dynet#1273]
In the document [here](http://dynet.readthedocs.io/en/latest/python_ref.html), the sample code below doesn't work anymore. I don't know how this mechanism works before adding `dynet_config` since it will init twice which is not allowed (https://github.com/clab/dynet/blob/master/dynet/init.cc#L220).

```
# Declare a DynetParams object
dyparams = dy.DynetParams()
# Fetch the command line arguments (optional)
dyparams.from_args()
# Set some parameters manualy (see the command line arguments documentation)
dyparams.set_mem(2048)
dyparams.set_random_seed(666)
# Initialize with the given parameters
dyparams.init() # or init_from_params(dyparams)
```

If no one uses this anymore, can we remove this both from code and document?

**Comment by pw2393** [comment]
The above code snippet works only when `import _dynet as dy`, rather than `import dynet as dy`.

`import dynet` itself inits the dynet; while `import _dynet` does not.  After the modification of global parameters, the dynet should have re-initialized. However, it now omits the re-initialization if dynet imported via `import dynet as dy`,  as shown:
```
WARNING: Attempting to initialize dynet twice. Ignoring duplicate initialization.
```
so the `dyparams.init()` fails to change the global parameters.

If re-initialization permits, `dy.DynetParams()` might be a more natural way for Python to configure dynet than `import dynet_config`.

**Comment by xunzhang** [comment]
@pw2393 You are right, `import _dynet` could work before `init()`, in [dynet.py](https://github.com/clab/dynet/blob/master/python/dynet.py.in#L15), but it's a little hacky since it cannot work together with external arguments passing. If we consider this as a more natural way to set arguments, I'll file a pull request enable this again(with external arguments passing) and update the related document. Thanks!

**Comment by yoavg** [comment]
The difference between `_dynet` and `dynet` is a leftover from when we had different modules for GPU and CPU (`_dynet` and `_gdynet`), and then dynet.py used to choose the correct one. Importing `_dynet` was never intended as something a user should do.

Another main difference is that `dynet` calls Init() as part of the loading process. I think this is more natural than requiring the use to call `Init()` by themselves each time they use dynet.

I suggest we remove `Init()`, `Init_from_params` and `DynetParams` from the python interface altogether, and allow only the `dynet_config` version.

Some things that are currently set in `Init`, like the random seed, or auto-batching, we can also allow to do post-init, by slightly adapting the C++ version.

**Comment by pw2393** [comment]
Thank you guys so much for the comments. 

Yea, users never import `_dynet` or `_gdynet`, and yea `dynet` init() with default global settings as part of importing process. I agree that it's natural not to require user to `init` by themselves, but probably not that ideal to use the current `dynet_config` to set globally:

> Suppose one has to import another module `dynet_config` BEFORE `dynet` to config –– which is the situation now –– it assumes that throughout the python program the global settings won't change. It's not bad, but for example it leads to a problem that when programmers wanna try dynet in jupyter notebook, they have to restart the whole kernel to reset the global params (because after `dynet` imported, `dynet_config` will not work; even if you try to import `dynet` again, `dynet_config` will still not work, due to the prevention of duplicate initialization).

post-init sounds nice.

**Comment by yoavg** [comment]
I totally agree with this. Because of the way the CPP core is structured, some values are set at loading time and are unchangeable. Most of these really can be re-set post init. The trickier ones are the memory pool sizes (though we should definitely be able to expose a `.resize()` method), and the devices initialization.

**Comment by xunzhang** [comment]
Cool! Let me summarize the previous discussion.

- The recommended way is to use `dynet_config`. For users who want to import twice in one place with different settings, please use `import _dynet` and do explicit init using `DynetParams` as a python wrapper workaround. So we will keep the old way for now.

- We'll add internal interfaces from c++ part to allow users to reset global settings, at least part of the arguments. This will be an enhancement so I think it will be tracked by another thread.

**Comment by yoavg** [comment]
I agree re moving point (2) to a different thread.

Re the first bullet, I am pretty sure initializing twice is *not possible* even with importing `_dynet`, because of the underlying backend. If it was, we wouldn't need `dynet_config` to begin with.

So unless you demonstrate this is not the case (i.e., double init is possible), I suggest removing `DynetParams`, OR integrating them into the `dynet_config` method, though I am not sure why this is needed.

**Comment by xunzhang** [comment]
My above first bullet is incorrect. So if that's the case, we don't need `DynetParams ` anymore, let's remove this, and add resetting feature in the backend later.
